examples/complex/jsondump.py

```diff
@@ -0,0 +1,245 @@
+        under the specified `path` in the `obj` dictionary.
+        """
+        for key in path[:-1]:
+            if not (key in obj and obj[key]):
+                return
+            obj = obj[key]
+        if path[-1] in obj and obj[path[-1]]:
+            obj[path[-1]] = func(obj[path[-1]])
+
+    @classmethod
+    def convert_to_strings(cls, obj):
+        """
+        Recursively convert all list/dict elements of type `bytes` into strings.
+        """
+        if isinstance(obj, dict):
+            return {cls.convert_to_strings(key): cls.convert_to_strings(value)
+                    for key, value in obj.items()}
+        elif isinstance(obj, list):
+            return [cls.convert_to_strings(element) for element in obj]
+        elif isinstance(obj, bytes):
+            return str(obj)[2:-1]
+        return obj
+
+    def dump(self, frame):
+        """
+        Transform and dump (write / send) a data frame.
+        """
+        for tfm in self.transformations:
+            for field in tfm['fields']:
+                self.transform_field(frame, field, tfm['func'])
+        frame = self.convert_to_strings(frame)
+
+        if self.outfile:
+            self.lock.acquire()
+            self.outfile.write(json.dumps(frame) + "\n")
+            self.lock.release()
+        else:
+            requests.post(self.url, json=frame, auth=(self.auth or None))
+
+    @staticmethod
+    def load(loader):
+        """
+        Extra options to be specified in `~/.mitmproxy/config.yaml`.
+        """
+        loader.add_option('dump_encodecontent', bool, False,
+                          'Encode content as base64.')
+        loader.add_option('dump_destination', str, 'jsondump.out',
+                          'Output destination: path to a file or URL.')
+        loader.add_option('dump_username', str, '',
+                          'Basic auth username for URL destinations.')
+        loader.add_option('dump_password', str, '',
+                          'Basic auth password for URL destinations.')
+
+    def configure(self, _):
+        """
+        Determine the destination type and path, initialize the output
+        transformation rules.
+        """
+        self.encode = ctx.options.dump_encodecontent
+
+        if ctx.options.dump_destination.startswith('http'):
+            self.outfile = None
+            self.url = ctx.options.dump_destination
+            ctx.log.info('Sending all data frames to %s' % self.url)
+            if ctx.options.dump_username and ctx.options.dump_password:
+                self.auth = (ctx.options.dump_username, ctx.options.dump_password)
+                ctx.log.info('HTTP Basic auth enabled.')
+        else:
+            self.outfile = open(ctx.options.dump_destination, 'a')
+            self.url = None
+            self.lock = Lock()
+            ctx.log.info('Writing all data frames to %s' % ctx.options.dump_destination)
+
+        self._init_transformations()
+
+    @concurrent
```

Comment:
Have you considered to just put messages into a single `queue.Queue` and have a dedicated thread for writing to disk? This way you don't need to spawn multiple threads per HTTP request.